lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+//   - Recursive case: If the block is not VALIDATED or VALIDATE_FAILED in the blockIndexByHash, we will perform all
+//     validations and add the block to the block index with the appropriate status (VALIDATED OR VALIDATE_FAILED) and
+//     return the new BlockNode.
+//   - Error case: Something goes wrong that doesn't result in the block being marked VALIDATE or VALIDATE_FAILED. In
+//     this case, we will add the block to the block index with status STORED and return the BlockNode.
+func (bc *Blockchain) validateAndIndexBlockPoS(block *MsgDeSoBlock) (*BlockNode, error) {
+	blockHash, err := block.Header.Hash()
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem hashing block %v", block)
+	}
+
+	// Base case - Check if the block is validated or validate failed. If so, we can return early.
+	blockNode, exists := bc.blockIndexByHash[*blockHash]
+	if exists && (blockNode.IsValidateFailed() || blockNode.IsValidated()) {
+		return blockNode, nil
+	}
+
+	// Run the validation for the parent and update the block index with the parent's status. We first
+	// check if the parent has a cached status. If so, we use the cached status. Otherwise, we run
+	// the full validation algorithm on it, then index it and use the result.
+	parentBlockNode, err := bc.validatePreviouslyIndexedBlockPoS(block.Header.PrevBlockHash)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem validating previously indexed block: ")
+	}
+
+	// Here's where it gets a little tricky. If the parent has a status of ValidateFailed, then we know we store
+	// this block as ValidateFailed. If the parent is not ValidateFailed, we ONLY store the block and move on.
+	// We don't want to store it as ValidateFailed because we don't know if it's actually invalid.
+	if parentBlockNode.IsValidateFailed() {
+		return bc.storeValidateFailedBlockWithWrappedError(block, errors.New("parent block is ValidateFailed"))
+	}
+
+	// If the parent block still has a Stored status, it means that we weren't able to validate it
+	// despite trying. The current block will also be stored as a Stored block.
+	if !parentBlockNode.IsValidated() {
+		return bc.storeBlockInBlockIndex(block)
+	}
+
+	// At this point, we know the parent block is validated. We can now perform all other validations
+	// on the current block. First we store the block in the block index.
+	blockNode, err = bc.storeBlockInBlockIndex(block)
+	if err != nil {
+		return nil, errors.Wrapf(err, "validateAndIndexBlockPoS: Problem adding block to block index: %v", err)
+	}
+
+	// Check if the block is properly formed and passes all basic validations.
+	if err = bc.isValidBlockPoS(block); err != nil {
+		return bc.storeValidateFailedBlockWithWrappedError(block, err)
+	}
+
+	// Validate the block's random seed signature
+	isValidRandomSeedSignature, err := bc.hasValidProposerRandomSeedSignaturePoS(block)
```

Comment:
Don't do this yet, but I'm just flagging that processing signatures is probably going to make historical block processing a lot slower. We don't need to fix it yet but put a TODO here, and wherever we do a signature check to indicate that we may want to flag this behind a `verifySignatures` check.